python/rikai/spark/sql/codegen/mlflow_registry.py

```diff
@@ -0,0 +1,303 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import os
+from typing import Any, Callable, Dict, IO, Mapping, Optional, Union
+from urllib.parse import urlparse
+
+try:
+    import mlflow
+except ImportError:
+    raise ImportError(
+        "Couldn't import mlflow. Please make sure to "
+        "`pip install mlflow` explicitly or install "
+        "the correct extras like `pip install rikai[mlflow]`"
+    )
+from mlflow.exceptions import MlflowException
+from mlflow.tracking import MlflowClient
+from pyspark.sql import SparkSession
+
+from rikai.conf import CONF_MLFLOW_TRACKING_URI
+from rikai.logging import logger
+from rikai.spark.sql.codegen.base import (
+    ModelSpec,
+    register_udf,
+    Registry,
+    udf_from_spec,
+)
+from rikai.spark.sql.exceptions import SpecError
+
+__all__ = ["MlflowRegistry"]
+
+
+class MlflowModelSpec(ModelSpec):
+    """Model Spec.
+
+    Parameters
+    ----------
+    run: mlflow.entities.Run
+        The Run object containing the spec info and ref to model artifact
+    tracking_uri: str
+        The mlflow tracking uri
+    options: Dict[str, Any], optional
+        Additionally options. If the same option exists in spec already,
+        it will be overridden.
+    validate: bool, default True.
+        Validate the spec during construction. Default ``True``.
+    """
+
+    def __init__(
+        self,
+        run: "mlflow.entities.Run",
+        tracking_uri: str,
+        options: Optional[Dict[str, Any]] = None,
+        validate: bool = True,
+    ):
+        self._run = run
+        spec = self._run_to_spec_dict()
+        spec.setdefault("options", {})
+        if options:
+            spec["options"].update(options)
+        super().__init__(spec, validate=validate)
+        self._artifact = None
+        self.tracking_uri = tracking_uri
+
+    @property
+    def run_id(self):
+        return self._run.info.run_id
+
+    @property
+    def start_time(self):
+        return self._run.info.start_time
+
+    @property
+    def end_time(self):
+        return self._run.info.end_time
+
+    def load_model(self) -> Any:
+        """Load the model artifact specified in this spec"""
+        old_uri = mlflow.get_tracking_uri()
+        try:
+            mlflow.set_tracking_uri(self.tracking_uri)
+            return getattr(mlflow, self.flavor).load_model(self.uri)
+        finally:
+            mlflow.set_tracking_uri(old_uri)
+
+    def _run_to_spec_dict(self):
+        """Convert the Run into a ModelSpec
+
+        Returns
+        -------
+        spec: Dict[str, Any]
+        """
+        tags = self._run.data.tags
+        spec = {
```

Comment:
do we expect these keys do not exist? if so, lets throw some actionable errors?

Reply:
yeah i can add a message about they should use rikai.mlflow.log_model, or add these to the options.